form.py

```diff
@@ -285,64 +310,319 @@ def open_config(file):
+    import requests
+    from bs4 import BeautifulSoup
+
+    form_id = "1FAIpQLSfWiBiihYkMJcZEAOE3POOKXDv6p4Ox4rX_ZRsQwu77aql8kQ"
+    url = to_form_url(form_id)
+
+    response = requests.get(url)
+    soup = BeautifulSoup(response.text, "html.parser")
+
+    info_soup = info_using_soup(soup)
+    info_json = info_using_json(form_json_data(soup))
+
+    for key in info_soup.keys() & info_json.keys():
+        assert info_soup[key] == info_json[key]
+
+# Create entries from info
+# `info` needs "types", "titles", "keys", "required", and "options"
+def entries_from_info(info):
+    entries = []
+    if info["takes_email"]:
+        args = (True, True, "extra", "emailAddress", "Email address", "")
+        entries.append(EntryInfo(*args))
+    for type, title, key, required, options in zip(
+        info["types"], info["titles"], info["keys"],
+        info["required"], info["options"],
+    ):
+        if options:
+            title = f"{title} ({', '.join(options)})"
+        entries.append(EntryInfo(required, True, type, key, title, ""))
+    return entries
+
+# Iterator of config lines from info
+def config_lines_from_info(info):
+    # First line should be a link that you can paste into a browser
+    yield to_normal_form_url(info["form_url"])
+
+    # Note that the file was auto-generated
+    yield f"# Auto-generated using form.py"
+
+    yield f"# {info['form_title']} - {info['form_description']}"
+    for entry in entries_from_info(info):
+        yield str(entry)
+
+parser = ArgumentParser(description="Automate Google Forms")
+parser.add_argument("target", default="config.txt", nargs="?",
+    help="file or url to use")
+modes = parser.add_mutually_exclusive_group()
+modes.add_argument("-p", "--process", action="store_true",
+    help="process the target config file and send the response")
+modes.add_argument("-c", "--convert", metavar="URL",
+    help="convert the form at the url into a config file at target")
+
+# Get and convert the form HTML
+def get_html_from_convert(convert):
+    if not convert.endswith(".URL"):
+        try:
+            with open(convert) as file:
+                return file.read()
+        except FileNotFoundError:
+            pass
 
-    if len(argv) == 2:
-        print(f"Using config file: {argv[1]}")
-        return argv[1]
+    # Used to get the form HTML
+    try:
+        import requests
+    except ImportError:
+        print("Form cannot be converted (missing requests library)")
+        sys.exit(3)
 
-    print("Using default name: config.txt")
-    return "config.txt"
+    # We're using to_form_url instead of to_normal_form_url. Apparently
+    # the -viewform URL doesn't have the form ready immediately but
+    # -formResponse does. Maybe its something with the page loading or
+    # some JS trickery.
+    url = to_form_url(convert)
+    response = requests.get(url)
+    return response.text
 
-# Returns config info of the passed file name
-def read_config(name):
+def main(args):
```

Comment:
This should be split into multiple functions (maybe `get_target_action`, `process`, and `convert`). However, this has already been done on the *better-parser* branch, so it could be ignored for now.

Reply:
Refactored in 4f71962e520b23e77bbb1639a48cf04d007f7dc2. The *better-parser* branch has similar functions but are also made for scripts, not just the command line.